src/main/java/org/dataone/hashstore/HashStore.java

```diff
@@ -73,66 +80,123 @@ public HashStore(Path storeDirectory)
             throw new NullPointerException("Invalid input stream, data is null.");
         }
-        if (pid == null || pid.isEmpty()) {
+        if (pid == null || pid.trim().isEmpty()) {
+            logHashStore.error("HashStore.storeObject - pid cannot be null or empty, pid: " + pid);
             throw new IllegalArgumentException("Pid cannot be null or empty, pid: " + pid);
         }
+        // Checksum cannot be empty or null if checksumAlgorithm is passed
+        if (checksumAlgorithm != null & checksum != null) {
+            if (checksum.trim().isEmpty()) {
+                logHashStore
+                        .error("HashStore.storeObject - checksum cannot be empty if checksumAlgorithm is supplied, pid: "
+                                + pid);
+                throw new IllegalArgumentException(
+                        "Checksum cannot be empty when checksumAlgorithm is supplied.");
+            }
+        }
+        // Cannot generate additional or checksum algorithm if it is not supported
+        if (additionalAlgorithm != null) {
+            boolean algorithmSupported = this.hashfs.isValidAlgorithm(additionalAlgorithm);
+            if (!algorithmSupported) {
+                logHashStore.error("HashStore.storeObject - additionalAlgorithm is not supported."
+                        + "additionalAlgorithm: " + additionalAlgorithm + ". pid: " + pid);
+                throw new IllegalArgumentException(
+                        "Additional algorithm not supported - unable to generate additional hex digest value. additionalAlgorithm: "
+                                + additionalAlgorithm + ". Supported algorithms: "
+                                + Arrays.toString(this.hashfs.supportedHashAlgorithms));
+            }
+        }
+        // Check support for checksumAlgorithm
+        if (checksumAlgorithm != null) {
+            boolean checksumAlgorithmSupported = this.hashfs.isValidAlgorithm(checksumAlgorithm);
+            if (!checksumAlgorithmSupported) {
+                logHashStore
+                        .error("HashStore.storeObject - checksumAlgorithm not supported, checksumAlgorithm: "
+                                + checksumAlgorithm + ". pid: " + pid);
+                throw new IllegalArgumentException(
+                        "Checksum algorithm not supported - cannot be used to validate object. checksumAlgorithm: "
+                                + checksumAlgorithm + ". Supported algorithms: "
+                                + Arrays.toString(this.hashfs.supportedHashAlgorithms));
+            }
+        }
 
         // Lock pid for thread safety, transaction control and atomic writing
         // A pid can only be stored once and only once, subsequent calls will
         // be accepted but will be rejected if pid hash object exists
         synchronized (objectLockedIds) {
             while (objectLockedIds.contains(pid)) {
```

Comment:
We don't need to store objects with the same identifier twice. So can we throw an exception if the objectLockedIds already contains the pid?

Reply:
I have updated the logic here to throw an exception if objectLockedIds already contains the pid (RunTimeException). When we create custom exceptions in HashStoreInterface, I will update this to a custom exception class that is more descriptive. Also discovered a bug in my thread tests and resolved it - thank you!